frontend/src/app/modules/users/users-selectors.js

```diff
@@ -1,14 +1,65 @@
 import { createSelector } from 'reselect'
 
+import * as Filters from '../../const/leaderboards-filters'
+import { didUserPlayMatch } from '../matches/matches-selectors'
+import {
+  computeLongestWinStreak,
+  computeWinRatio,
+} from '../matches/matches-computations'
+
 export const getUsers = state => state.users
+const getMatches = state => state.matches
+const getFilters = state => state.filters
 
 export const getUser = createSelector(
   getUsers,
   (stats, userId) => userId,
   (users, userId) => users.find(user => user.id === userId),
 )
 
-export const getTopUsers = createSelector(
+export const getTopRatedUsers = createSelector(
   getUsers,
   users => [...users].sort((user1, user2) => user2.rating - user1.rating),
 )
+
+export const getTopUsers = createSelector(
+  getUsers,
+  getMatches,
+  getFilters,
+  (users, matches, filters) => sortTopUsers([...users], [...matches], filters),
+)
+
+const sortTopUsers = (users, matches, filters) => {
+  const order = filters.order === Filters.orderTypes.ASC ? -1 : 1
+  const minDate = new Date()
+  minDate.setDate(minDate.getDate() - filters.timespan)
+  const matchesLast = matches
+    .filter(match => filters.timespan === Filters.timespanTypes.AllTime || match.date > minDate)
+    .sort((match1, match2) => match2.date - match1.date)
+
+  users = users.map(user => {
```

Comment:
if we want to be superclear about this, we could do `const usersWithStatistics = users.map()`

Reply:
actually you could even extract the whole thing into a separate function. enhanceUsersWithStatistics

Reply:
@novellizator +1, it's probably better not to reassign the function argument.

Reply:
done [here](https://github.com/salsita/foosball-rating/pull/92/commits/d7c17da31ec421acbb90f4a19ce22c34412cdf35#diff-6bb144976a04c9e3de620e0c0828e32e)